keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+      meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Automatic certificates copy makes easier to create HA clusters with the kubeadm tool using exactly
+the same `kubeadm init` and `kubeadm join` commands the users are familiar with.
+
+## Motivation
+
+As confirmed by the recent [kubeadm survey](https://drive.google.com/file/d/1eN9sGsdXWpurmplbEVn9UX5NiseQqlzO/view?usp=sharing),
+support for high availability cluster is one of the most requested features for kubeadm.
+
+A lot of effort was already done in kubeadm for achieving this goal, among them the redesign
+of the kubeadm config file and its graduation to beta and the implementation of the
+[`kubeadm join --control-plane workflow (KEP0015)`](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0015-kubeadm-join-control-plane.md),
+but the solution currently in place stills requires the manual copy of cluster certificates from
+the bootstrap control-plane node to secondary control-plane nodes.
+
+This KEP introduces automatic certificates copy, eliminating the manual operation described
+above and completing the kubeadm solution for creating HA clusters.
+
+### Goals
+
+- Usability: use exactly the same `kubeadm init` and `kubeadm join` commands the users are
+  familiar with.
+- Security: provide a solution that enables the secure copy of cluster certificates between
+  control-plane nodes
+
+## Proposal
+
+### Constraints
+
+The solution described in this proposal is deeply influenced by following constraints:
+
+1. Kubeadm can execute actions only on the machine where it is running e.g. it is not
+   possible to execute actions on other nodes.
+2. `kubeadm init` and `kubeadm join` are separated actions (executed on separated machines/at
+   different times).
+3. During the join workflow, kubeadm can access the cluster only using identities with
+   limited grants, namely `system:unauthenticated` or `system:node-bootstrapper`.
+
+### Key element of the proposal
+
+At a high level, the proposed solution can be summarized by the following workflow:
+
+**On control plane node 1:**
+
+```bash
+kubeadm init --upload-certs
+```
+
+The new `--upload-certs` flag will trigger a new kubeadm init phase executing following
+actions:
+
+1. An encryption key will be generated; such **key will never be stored on cluster**
+
+2. Cluster certificates will be encrypted using the above key
+
+3. Encrypted certificates will be stored in a new Kubernetes secret named `kubeadm-certs`;
+   it is important to notice that:
+
+   - This secret is the technical solution that provides a temporary bridge between
+     `kubeadm init` and `kubeadm join` commands/between different nodes.
+
+   - Without the encryption key, this secret contains a harmless blog of bytes.
+
+4. A second bootstrap token will be generated, with a shorter duration than the
+   join token (the TTL-token)
```

Comment:
should the TTL be an implementation detail outside of the scope of the KEP?

either way, my vote goes for 1h.
non-customizable.